advertise/addr.go

```diff
@@ -13,91 +14,152 @@ import (
 // local network traffic for Unix-like platforms.
 var DefaultInterfaces = []string{"eth0", "en0"}
 
-// FirstAddress returns the first IPv4 address from the given interface names.
-// Addresses used for APIPA will be ignored if possible.
-func FirstAddress(interfaces []string) (net.IP, error) {
+// FirstAddress returns the "best" IP address from the given interface names.
+// "best" is defined as follow in decreasing order:
+// - IPv4 valid and not link-local unicast
+// - IPv6 valid and not link-local unicast
+// - IPv4 valid and link-local unicast
+// - IPv6 valid and link-local unicast
+// If none of the above are found, an invalid address is returned.
+// Loopback addresses are never selected.
+// If no interfaces are provided, all of the system's network interfaces will retrieved via net.Interfaces
+// and used to find the "best" ip address.
+func FirstAddress(interfaces []string) (string, error) {
+	return firstAddress(interfaces, getInterfaceAddresses, net.Interfaces)
+}
+
+// NetworkInterfaceAddressGetter matches the signature of net.InterfaceByName() to allow for test mocks.
+type NetworkInterfaceAddressGetter func(name string) ([]netip.Addr, error)
+
+// InterfaceLister matches the signature of net.Interfaces() to allow for test mocks.
+type InterfaceLister func() ([]net.Interface, error)
```

Comment:
These types are only used in unit tests so they shouldn't be publicly exposed as part of our public API.